src/irmin/s.ml

```diff
@@ -137,6 +170,25 @@ module type APPEND_ONLY_STORE_MAKER = functor (K : Type.S) (V : Type.S) -> sig
   include CLOSEABLE with type 'a t := 'a t
 end
 
+module type UNSAFE_APPEND_ONLY_STORE = sig
+  include APPEND_ONLY_STORE
+
+  val add : 'a t -> key -> value -> unit Lwt.t
+  val unsafe_mem : 'a t -> key -> bool
+  val unsafe_find : 'a t -> key -> value option
+  val unsafe_append : 'a t -> key -> value -> unit
```

Comment:
These `unsafe_` functions are a bit misleading: there are not unsafe, they just don't use `Lwt` :-)

I'm super curious to know how hard it would be to remove that module completely and modify `Irmin.Inode` to use `lwt` as well. Last time I tried it was super painful, but now that the split is much saner maybe that's easier.

Reply:
Is the fact that they take a `key` the reason why they are called unsafe? I.e. if the key doesn't match the value this is undefined

Reply:
ha yes that's true ; but it's a bit odd here that they both take a key AND they don't use lwt.

Reply:
I agree this is super painful ; 

but the whole point of these `Make_ext` functor is to be able to change backend easily - all other backends are using `Lwt`right now, so if we want an in-memory version of inodes, or a filesystem ones, etc. we'll have to re-implement a non-lwt implementation version of all of these which is also not super great.

So the balance here is: either we make the inode functor looks more "normal", or we rewrite all the backend in direct style. In the long term (with effects) the later is probably what is best. In the short term, either we are fine with only an in-memory direct-style backend, or we bite the bullet and rewrite the inode functor :-)

Reply:
I'm giving it a try!